EventFilter/CastorRawToDigi/src/ZdcUnpacker.cc

```diff
@@ -0,0 +1,330 @@
+    if (!dccHeader->getSpigotPresent(spigot)) continue;
+
+    int retval=dccHeader->getSpigotData(spigot,htr,raw.size());
+    if (retval!=0) {
+      if (retval==-1) {
+        if (!silent) edm::LogWarning("Invalid Data") << "Invalid HTR data (data beyond payload size) observed on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId();
+        report.countSpigotFormatError();
+      }
+      continue;
+    }
+    // check
+    if (dccHeader->getSpigotCRCError(spigot)) {
+      if (!silent)
+        edm::LogWarning("Invalid Data") << "CRC Error on HTR data observed on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId();
+      report.countSpigotFormatError();
+      continue;
+    }
+    if (!htr.check()) {
+      if (!silent)
+        edm::LogWarning("Invalid Data") << "Invalid HTR data observed on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId();
+      report.countSpigotFormatError();
+      continue;
+    }
+    if (htr.isHistogramEvent()) {
+      if (!silent)
+        edm::LogWarning("Invalid Data") << "Histogram data passed to non-histogram unpacker on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId();
+      continue;
+
+    }
+    if ((htr.getFirmwareFlavor()&0xE0)==0x80) { // some kind of TTP data
+      if (colls.ttp!=0) {
+        HcalTTPUnpacker ttpUnpack;
+        colls.ttp->push_back(HcalTTPDigi());
+        ttpUnpack.unpack(htr,colls.ttp->back());
+      } else {
+        LogDebug("ZdcUnpackerHcalTechTrigProcessor") << "Skipping data on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId() << " which is from the TechTrigProcessor (use separate unpacker!)";
+      }
+      continue;
+    }
+    if (htr.getFirmwareFlavor()>=0x80) {
+      if (!silent) edm::LogWarning("ZdcUnpacker") << "Skipping data on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId() << " which is of unknown flavor " << htr.getFirmwareFlavor();
+      continue;
+    }
+    // calculate "real" number of presamples
+    //int nps=htr.getNPS()-startSample_;
+
+    // get pointers
+    htr.dataPointers(&daq_first,&daq_last,&tp_first,&tp_last);
+    unsigned int smid=htr.getSubmodule();
+    int htr_tb=smid&0x1;
+    int htr_slot=(smid>>1)&0x1F;
+    int htr_cr=(smid>>6)&0x1F;
+
+    //////////////////////////////////////////////
+    qie_begin=(HcalQIESample*)daq_first;
+    qie_end=(HcalQIESample*)(daq_last+1); // one beyond last..
+
+    for (qie_work=qie_begin; qie_work!=qie_end;) {    
+      if (qie_work->raw()==0xFFFF) {
+        qie_work++;
+        continue; // filler word
+      }
+      // lookup the right channel
+   CastorElectronicsId eid(qie_work->fiberChan(),qie_work->fiber(),spigot,dccid);
+      eid.setHTR(htr_cr,htr_slot,htr_tb);
+
+      eid.setHTR(htr_cr,htr_slot,htr_tb);
+
+      if (qie_work->fiberChan()==0 && qie_work->fiber()==1 && spigot==0 && htr_tb==1)
```

Comment:
Hi Slava,

I've just fixed this issue, it will be apparent in the next update. I've also fixed the const cast issue you brought up. Sorry for the pre-mature commit!  I didn't realize it would automatically set off the next round of checks.

At this point our major issue is adding capability for running both old and new data, which we're still stuck on.

Chris

> On Nov 18, 2015, at 19:54, Slava Krutelyov <notifications@github.com> wrote:
> 
> In EventFilter/CastorRawToDigi/src/ZdcUnpacker.cc:
> 
> > +    //////////////////////////////////////////////
> > +    qie_begin=(HcalQIESample*)daq_first;
> > +    qie_end=(HcalQIESample*)(daq_last+1); // one beyond last..
> > +
> > +    for (qie_work=qie_begin; qie_work!=qie_end;) {    
> > +      if (qie_work->raw()==0xFFFF) {
> > +        qie_work++;
> > +        continue; // filler word
> > +      }
> > +      // lookup the right channel
> > +   CastorElectronicsId eid(qie_work->fiberChan(),qie_work->fiber(),spigot,dccid);
> > +      eid.setHTR(htr_cr,htr_slot,htr_tb);
> > +
> > +      eid.setHTR(htr_cr,htr_slot,htr_tb);
> > +
> > +      if (qie_work->fiberChan()==0 && qie_work->fiber()==1 && spigot==0 && htr_tb==1)
> 
> the following ~140 lines can be compacted into 10.
> What's the reason for this massive copy-pasted slew of "else if (something){ sameCode; }" ?
> 
> —
> Reply to this email directly or view it on GitHub.
>